[FIX] IsotopeDistribution bug for gapped isotopes (e.g. Bromine)

cbielow opened the pull request:
Elemental isotope distributions can be gapped, i.e. Bromine only has Br79 and Br81 as stable isotopes.

Our convolution algorithm however, assumes ungapped isotope distributions, i.e. an explicit Br80=0 (which is missing in our case).

This PR fills the required gaps before computing the isotope distribution.

timosachsenberg commented:
+1

cbielow commented:
test still fails.. will fix

cbielow commented:
@hroest: can you please confirm that the minor score deviation for DIAscoring is expected.

hroest commented on src/tests/class_tests/openms/source/IsotopeDistribution_test.cpp:
```diff
@@ -161,6 +161,39 @@ START_SECTION(IsotopeDistribution& operator *= (Size factor))
+    container.push_back(make_pair<Size, double>(159, 0.0));
+    container.push_back(make_pair<Size, double>(160, 0.49990478));
+    container.push_back(make_pair<Size, double>(161, 0.0));
+    container.push_back(make_pair<Size, double>(162, 0.24314761));
```
could you please add some data where you got these numbers, how you calculated the "true" numbers here so that future generations can easily figure out what you did here?

hroest commented on src/tests/class_tests/openms/source/IsotopeDistribution_test.cpp:
```diff
@@ -161,6 +161,39 @@ START_SECTION(IsotopeDistribution& operator *= (Size factor))
+    container.push_back(make_pair<Size, double>(173, 0.005348981146));
+    container.push_back(make_pair<Size, double>(174, 0.240545930573));
+    container.push_back(make_pair<Size, double>(175, 0.002601679427));
+    for (Size i = 0; i != id.size(); ++i)
```
same here

hroest commented:
:+1:  - can you add some explanation / Python code to the ?

hroest commented:
@cbielow it is expected as these scores rely on the isotopic distribution. If the calculation of the distribution is now more correct then yes, this is probably what we want. The changes are quite minor, about 20 ppm in the actual numbers so I dont think that is a problem

cbielow commented:
rebased to enable merge again.

added some comments to test code (even though I would expect people to understand convolution if they work on isotope distributions?!?)

cbielow commented:
@hroest my question regarding the DIA was more about the reasoning why the DIA score would change if gapped isotopes are now correct, i.e. why would DIA look at Bromine and other gapped isotopes to calculate a score?!

timosachsenberg commented:
can you easily check in CDASH if this change might be unrelated to this PR?

hroest commented:
no, I dont see why there should be a change if there is only an improvement for gapped isotopes. This code is located somewhere in

OpenMS/ANALYSIS/OPENSWATH/DIAPrescoring.h

and calls `addIsotopes2Spec` and `getAveragineIsotopeDistribution` in ./source/ANALYSIS/OPENSWATH/DIAHelper.cpp 

given that the problems only occur in this part of the code and not in any other code dealing with isotopes seems to indicate that the problem is only located here. This part of the code is written by Witold and maybe witold could shed some light onto this. But given that the differences are minor, I think we should be fine

timosachsenberg commented:
so should we merge?
